prov/rxm/src/rxm_ep.c

```diff
@@ -2437,11 +2435,41 @@ int rxm_endpoint(struct fid_domain *domain, struct fi_info *info,
 	(*ep_fid)->fid.ops = &rxm_ep_fi_ops;
 	(*ep_fid)->ops = &rxm_ops_ep;
 	(*ep_fid)->cm = &rxm_ops_cm;
-	(*ep_fid)->msg = &rxm_ops_msg;
-	(*ep_fid)->tagged = &rxm_ops_tagged;
 	(*ep_fid)->rma = &rxm_ops_rma;
 
+	(*ep_fid)->msg = calloc(1, sizeof(*(*ep_fid)->msg));
+	if (!(*ep_fid)->msg) {
+		ret = -FI_ENOMEM;
+		goto err4;
+	}
+	*(*ep_fid)->msg = rxm_ops_msg;
+
+ 	(*ep_fid)->tagged = calloc(1, sizeof(*(*ep_fid)->tagged));
+	if (!(*ep_fid)->tagged) {
+		ret = -FI_ENOMEM;
+		goto err5;
+	}
+	*(*ep_fid)->tagged = rxm_ops_tagged;
+
+	if (rxm_ep->util_ep.domain->threading != FI_THREAD_SAFE) {
+		(*ep_fid)->msg->inject = rxm_ep_inject_fast;
+		(*ep_fid)->msg->injectdata = rxm_ep_injectdata_fast;
+		(*ep_fid)->tagged->inject = rxm_ep_tinject_fast;
+		(*ep_fid)->tagged->injectdata = rxm_ep_tinjectdata_fast;
+	}
+
+	if (rxm_ep_get_inject_buf_limit(rxm_ep) == 0) {
+		(*ep_fid)->msg->inject = fi_no_msg_inject;
+		(*ep_fid)->msg->injectdata = fi_no_msg_injectdata;
+		(*ep_fid)->tagged->inject = fi_no_tagged_inject;
+		(*ep_fid)->tagged->injectdata = fi_no_tagged_injectdata;
+	}
```

Comment:
Can you just have 4 static structs instead of dynamically allocating the structures per ep?  I.e. msg_ops_fast, msg_ops_slow, tagged_ops_fast, tagged_ops_slow?  Only use better names.  :)

Reply:
Sean, we should have tagged/msg fast/slow and for each of them - w/ and w/o inject function. When MSG provider has inject size <= sizeof(struct rxm_hdr), it would be great to report -FI_ENOSYS if user calss the inject function. For example, MPICH/CH4 does fi_tinject if a size of provided buffer is <= inject size. It doesn't work correct if provider reports inject size = 0.  
Whatyou think about to report -FI_ENOSYS?

Reply:
So, there would be 8 possible combinations in total?  If that's the case, dynamically allocating the ops structure makes sense.  The app shouldn't be opening that many endpoints.

If the supportable inject size is 0, then setting the call to one of the 'no inject' functions to return enosys makes sense to me.

Reply:
@shefty 

> Can you just have 4 static structs instead of dynamically allocating the structures per ep?

Done, we don't need `fi_no_[t]inject*` anymore